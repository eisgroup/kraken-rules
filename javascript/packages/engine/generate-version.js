const fs = require('fs')
const path = require('path')
const xmlParser = require('fast-xml-parser')

console.log('----------------------------------');
console.log('generating engine version constant');
console.log('----------------------------------');


const pomPath = path.join(
    __dirname,
    '../',
    '../',
    '../',
    'pom.xml'
)

console.log(`  reading '${pomPath}'`);

const pom = xmlParser.parse(
    fs.readFileSync(
        pomPath,
        { encoding: 'utf8' }
    )
)

console.log(`  version is: '${pom.project.version}'`);

const genFolderPath = path.join(
    __dirname,
    'src',
    'engine',
    'generated'
)

if (!fs.existsSync(genFolderPath)) {
    fs.mkdirSync(genFolderPath)
}

const genPath = path.join(
    __dirname,
    'src',
    'engine',
    'generated',
    'engineVersion.ts'
);

console.log(`  writing to path: '${genPath}'`);

fs.writeFileSync(
    genPath,
    `
// generated code    
// do not change it manually    
export const ENGINE_VERSION = "${pom.project.version}";

`, { encoding: 'utf8' }
)

console.log('  backend version is written');

process.exit(0)