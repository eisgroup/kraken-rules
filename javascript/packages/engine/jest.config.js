module.exports = {
    moduleFileExtensions: ['ts', 'tsx', 'js'],
    transform: {
        '^.+\\.ts$': ['@swc/jest'],
    },
    testMatch: ['**/?(*.)test.ts'],
    verbose: true,
    cacheDirectory: './target/tmp/',
    coverageDirectory: './target/coverage',
    coverageThreshold: {
        global: {
            branches: 93,
            functions: 96,
            lines: 96,
            statements: 96,
        },
    },
    collectCoverageFrom: [
        '**/*[^d\\.].ts',
        '!**/node_modules/**',
        '!**/target/**',
        '!**/__tests__/**',
        '!**/index.ts',
        '!**/**Error**.ts',
        '!**/Reducer.ts', //interface only
        '!**/RulePayloadHandler.ts', //interface only
        '!**/RepositoryClient.ts', //deprecated
        '!**/info/iterators/**', //deprecated
        '!**/config.ts',
        '!**/DevelopmentLogger.ts',
    ],
    testEnvironment: 'node',
    setupFilesAfterEnv: ['./__tests__/console.setup.js', './__tests__/matchers.setup.ts'],
}
