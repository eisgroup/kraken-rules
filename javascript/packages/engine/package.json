{
    "name": "kraken-typescript-engine",
    "version": "1.55.1-stage.0",
    "description": "",
    "main": "target/dist/js/src/index.js",
    "types": "target/dist/definitions/src/index.d.ts",
    "author": "Pavel Surinin <psurinin@eisgroup.com> (http://eisgroup.com)",
    "license": "Apache-2.0",
    "publishConfig": {
        "registry": "https://genesis-npm-release.exigengroup.com/repository/genesis-npm-release/"
    },
    "typedocMain": "src/index.ts",
    "dependencies": {
        "bignumber.js": "9.0.1",
        "declarative-js": "3.8.15",
        "fast-deep-equal": "2.0.1",
        "kraken-engine-api": "1.55.1-stage.0",
        "kraken-model": "1.55.1-stage.0",
        "moment": "2.29.4"
    },
    "devDependencies": {
        "@types/jest": "27.5.1",
        "@types/lodash": "4.14.116",
        "@types/node": "13.9.0",
        "child_process": "1.0.2",
        "fast-xml-parser": "^3.19.0",
        "jest": "27.5.1",
        "jsdom": "11.3.0",
        "jsdom-global": "3.0.2",
        "kraken-jest-matchers": "1.55.1-stage.0",
        "kraken-model-builder": "1.55.1-stage.0",
        "kraken-test-product": "1.55.1-stage.0",
        "kraken-test-product-model-tree": "1.55.1-stage.0",
        "prettier": "1.14.3",
        "rimraf": "3.0.2",
        "ts-node": "4.1.0",
        "typescript": "5.1.6"
    },
    "scripts": {
        "ci-versions": "echo \"typescript version\" && tsc -v && echo \"jest version\" && jest -v",
        "test": "jest -c jest.config.js",
        "test:coverage": "jest -c jest.config.js --coverage",
        "lint": "eslint src __tests__",
        "compile": "node generate-version && tsc --pretty",
        "postcompile": "yarn transpile",
        "transpile": "echo 'no transpile'",
        "del:compiled": "rimraf target",
        "build": "yarn run compile && yarn run lint && yarn run test:coverage",
        "prepare": "yarn run ci-versions && yarn run del:compiled && yarn run compile && yarn run lint -- --fix && yarn run test:coverage",
        "release": "yarn publish"
    },
    "files": [
        "target/dist/definitions",
        "target/dist/js"
    ],
    "gitHead": "efcf1a5e805088ee8af67ef21c2b8ce2fea62eb4"
}