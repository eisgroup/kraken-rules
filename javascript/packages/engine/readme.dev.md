# Kraken Typescript Engine

## Tests

To run tests enter command in console:
`npm test`
This command will run all tests without coverage

To run tests with coverage enter command in console:
`npm run test:coverage`

Kraken engine have its own matchers in jest, that can be used `expect(value).<matcher>`

```typescript
expect(results: EntryPointResult).k_toHaveExpressionsFailures(num?: number): void;
expect(result: RuleEvaluationResult).k_toBeValidRuleResult(): void;
expect(date: Date).k_toBeDate(argument: Date): void;
expect(date: Date).k_toBeDateTime(argument: Date): void;
expect(date: Date).k_toBeTodayDate(): void;
expect(results: EntryPointResult).k_toMatchResultsSnapshots();
/**
 * Matches rule results according to provided parameters, 
 * and check to haven no expression failures. For assert it uses
 * '.k_toHaveExpressionsFailures(0)'
 * @param stats     statistics results should match
 */
expect(results: EntryPointResult).k_toMatchResultsStats(stats: {
    total: number,
    critical?: number,
    warning?: number,
    info?: number,
    disabled?: number,
    hidden?: number
});
```

## Lint
To lint project enter command in console:
`npm run lint`

TSLint can fix some errors. To do this enter command in console:
`npm run lint -- --fix`

## Committing
Before commiting run this command:
`npm run precommit` this will fix some linter issues if they are present, run linter, 
tests with coverage and compile typescript to javascript
