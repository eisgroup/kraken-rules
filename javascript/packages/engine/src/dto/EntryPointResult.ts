/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { FieldEvaluationResult } from "./FieldEvaluationResult";
import { RuleEvaluationResults } from "./RuleEvaluationResults";
import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult;

import { Reducer } from "declarative-js";
import flat = Reducer.flat;
import { ConditionEvaluationResult } from "./ConditionEvaluationResult";

/**
 * Contains evaluation results from specific entry point
 */
export class EntryPointResult {
    private readonly results: Record<string, FieldEvaluationResult>;
    readonly evaluationTimestamp: Date;
    constructor(results: Record<string, FieldEvaluationResult>) {
        this.results = results;
        this.getApplicableResults.bind(this);
        this.evaluationTimestamp = new Date();
    }
    /**
     * Flatten contents into single list of {@link RuleEvaluationResult}s
     */
    getAllRuleResults(): RuleEvaluationResult[] {
        return Object.values(this.results)
            .map(x => x.ruleResults)
            .reduce(flat, []);
    }
    getFieldResults(): Record<string, FieldEvaluationResult> {
        return this.results;
    }

    getApplicableResults(): RuleEvaluationResult[] {
        return this.getAllRuleResults()
            .filter(r => ConditionEvaluationResult.isApplicable(r.conditionEvaluationResult));
    }

    static empty(): EntryPointResult {
        return new EntryPointResult({});
    }
}
