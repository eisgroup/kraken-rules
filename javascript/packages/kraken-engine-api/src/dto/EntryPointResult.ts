/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { FieldEvaluationResult } from './FieldEvaluationResult'
import { RuleEvaluationResults } from './RuleEvaluationResults'
import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult

/**
 * Contains evaluation results from specific entry point
 */
export interface EntryPointResult {
    readonly evaluationTimestamp: Date
    readonly ruleTimeZoneId: string

    /**
     * Returns entry point result mapped by field name.
     */
    getFieldResults(): Record<string, FieldEvaluationResult>

    /**
     * Flattens and returns field results into an array of {@link RuleEvaluationResult}
     */
    getAllRuleResults(): RuleEvaluationResult[]

    /**
     * Flattens and returns field results into an array of {@link RuleEvaluationResult}
     * filtered by rule condition evaluation status.
     */
    getApplicableResults(): RuleEvaluationResult[]
}
