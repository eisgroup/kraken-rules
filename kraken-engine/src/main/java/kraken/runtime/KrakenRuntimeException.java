/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package kraken.runtime;

import kraken.exception.KrakenException;
import kraken.message.SystemMessage;

/**
 * Superclass for all runtime execution exceptions
 *
 * @author rimas
 * @since 1.0
 */
public class KrakenRuntimeException extends KrakenException {

    private static final long serialVersionUID = -7921177886566471653L;

    /**
     * Constructor to build exception with message
     *
     * @param message   message text
     */
    public KrakenRuntimeException(SystemMessage message) {
        super(message);
    }

    /**
     * Constructor to build exception instance with message and cause
     *
     * @param message   message text
     * @param cause     cause exception
     */
    public KrakenRuntimeException(SystemMessage message, Throwable cause) {
        super(message, cause);
    }

}
