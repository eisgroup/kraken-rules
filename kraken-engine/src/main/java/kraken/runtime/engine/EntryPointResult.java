package kraken.runtime.engine;

import kraken.annotations.API;
import kraken.runtime.engine.dto.FieldEvaluationResult;
import kraken.runtime.engine.dto.RuleEvaluationResult;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Contains evaluation results from specific entry point
 *
 * @author rimas
 * @since 1.0
 */
@API
public class EntryPointResult {

    /**
     * Field evaluation results, mapped to field id string
     */
    private Map<String, FieldEvaluationResult> fieldResults;

    private transient LocalDateTime evaluationTimeStamp;
    /**
     * Default constructor
     */
    public EntryPointResult() {
        fieldResults = new HashMap<>();
        evaluationTimeStamp = LocalDateTime.now();
    }

    public EntryPointResult(Map<String, FieldEvaluationResult> fieldResults, LocalDateTime evaluationTimeStamp) {
        this.fieldResults = new HashMap<>(fieldResults);
        this.evaluationTimeStamp = evaluationTimeStamp;
    }

    /**
     * Flatten contents into single list of {@link RuleEvaluationResult}s
     *
     * @return  list of all contained {@link RuleEvaluationResult}s
     */
    public List<RuleEvaluationResult> getAllRuleResults() {
        return fieldResults.entrySet().stream()
                .flatMap(e -> e.getValue().getRuleResults().stream())
                .collect(Collectors.toList());
    }

    public List<RuleEvaluationResult> getApplicableRuleResults() {
        return fieldResults.entrySet().stream()
                .flatMap(e -> e.getValue().getRuleResults().stream())
                .filter(r -> r.getConditionEvaluationResult().isApplicable())
                .collect(Collectors.toList());
    }

    public LocalDateTime getEvaluationTimeStamp() {
        return evaluationTimeStamp;
    }

    public Map<String, FieldEvaluationResult> getFieldResults() {
        return fieldResults;
    }

}
