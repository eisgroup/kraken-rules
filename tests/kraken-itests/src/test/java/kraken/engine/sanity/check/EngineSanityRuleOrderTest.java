package kraken.engine.sanity.check;

import kraken.runtime.KrakenRuntimeException;
import kraken.runtime.engine.EntryPointResult;
import kraken.testproduct.domain.*;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Ignore;
import org.junit.Test;

import java.math.BigDecimal;
import java.util.List;

import static io.github.jsonSnapshot.SnapshotMatcher.start;
import static io.github.jsonSnapshot.SnapshotMatcher.validateSnapshots;
import static kraken.testing.matchers.KrakenMatchers.*;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static kraken.test.KrakenItestMatchers.matchesSnapshot;

/**
 * @author psurinin@eisgroup.com
 */
public class EngineSanityRuleOrderTest extends SanityEngineBaseTest {

    @BeforeClass
    public static void beforeAll() {
        start();
    }

    @AfterClass
    public static void afterAll() {
        validateSnapshots();
    }

    @Test
    public void shouldValidateEvaluateRulesInOrder() {
        final Policy policy = new Policy();
        Vehicle vehicle = new Vehicle();
        COLLCoverage coverage = new COLLCoverage();
        vehicle.setCollCoverages(List.of(coverage));
        policy.setRiskItems(List.of(vehicle));
        Insured insured = new Insured();
        insured.setHaveChildren(true);
        policy.setInsured(insured);
        BillingInfo billingInfo = new BillingInfo();
        CreditCardInfo creditCardInfo = new CreditCardInfo();
        creditCardInfo.setBillingAddress(new BillingAddress());
        billingInfo.setCreditCardInfo(creditCardInfo);
        policy.setBillingInfo(billingInfo);
        policy.setParties(List.of(new Party()));
        final EntryPointResult result = engine.evaluate(policy, "RuleOrder");

        assertThat(coverage.getCode(), is("PartyPartyAddress"));

        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasApplicableResults(7));
        assertThat(result, hasNoValidationFailures());
        assertThat(result, matchesSnapshot());
    }

    @Test
    public void shouldValidateRulesWithComplexFieldExpressionsInOrder() {
        final Policy policy = new Policy();
        BillingInfo billingInfo = new BillingInfo();
        billingInfo.setCreditCardInfo(new CreditCardInfo());
        policy.setBillingInfo(billingInfo);

        //RRCoverage
        Vehicle vehicle = new Vehicle();
        RRCoverage rrCoverage = new RRCoverage();
        vehicle.setRentalCoverage(rrCoverage);
        policy.setRiskItems(List.of(vehicle));

        //Insured
        Insured insured = new Insured();
        BillingAddress billingAddress = new BillingAddress();
        insured.setAddressInfo(billingAddress);
        policy.setInsured(insured);

        //Party
        Party party = new Party();
        PartyRole partyRole = new PartyRole();
        party.setRoles(List.of(partyRole));
        policy.setParties(List.of(party));

        final EntryPointResult result = engine.evaluate(policy, "RuleOrderWithComplexField");

        assertThat(policy.getPolicyNumber(), is("doNotSolicit is true"));
        assertThat(insured.getAddressInfo().getDoNotSolicit(), is(true));
        assertThat(rrCoverage.getLimitAmount(), is(new BigDecimal(10)));
        assertThat(billingAddress.getCity(), is("San Diego"));
        assertThat(partyRole.getLimit(), is(new BigDecimal(100)));


        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasApplicableResults(6));
        assertThat(result, hasNoValidationFailures());
        assertThat(result, matchesSnapshot());
    }

    @Test
    public void shouldValidateEvaluateDefaultRulesInOrder() {
        final Policy policy = new Policy();
        CarCoverage carCoverage = new CarCoverage();
        policy.setCoverage(carCoverage);

        TransactionDetails transactionDetails = new TransactionDetails();
        policy.setTransactionDetails(transactionDetails);

        final EntryPointResult result = engine.evaluate(policy, "FunctionCheck-RulesUsingFunctionOrderCheck");

        assertThat(carCoverage.getLimitAmount(), is(new BigDecimal("1000")));
        assertThat(transactionDetails.getTotalLimit(), is(new BigDecimal("1000")));

        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasApplicableResults(2));
        assertThat(result, hasNoValidationFailures());
        assertThat(result, matchesSnapshot());
    }

    @Test
    public void shouldEvaluateRuleOrderWithNoCycleInFunction() {
        Vehicle vehicle = new Vehicle();
        RRCoverage rrCoverage = new RRCoverage();
        vehicle.setRentalCoverage(rrCoverage);
        COLLCoverage collCoverage1 = new COLLCoverage(new BigDecimal("50"));
        COLLCoverage collCoverage2 = new COLLCoverage(new BigDecimal("75"));
        vehicle.setCollCoverages(List.of(collCoverage1, collCoverage2));
        Policy policy = new Policy();
        policy.setRiskItems(List.of(vehicle));

        EntryPointResult result = engine.evaluate(policy, "RuleOrderWithNoCycleInFunction");

        assertThat(policy.getPolicyNumber(), is("Large"));
        assertThat(rrCoverage.getLimitAmount(), is(new BigDecimal("125")));

        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasApplicableResults(2));
        assertThat(result, hasNoValidationFailures());
    }

}
