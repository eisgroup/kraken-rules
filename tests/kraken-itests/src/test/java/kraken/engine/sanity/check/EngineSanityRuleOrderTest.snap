kraken.engine.sanity.check.EngineSanityRuleOrderTest.shouldValidateEvaluateDefaultRulesInOrder=[
  {
    "CarCoverage:-1:limitAmount": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "CarCoverage",
        "fieldName": "limitAmount",
        "fieldPath": "limitAmount"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "Default Car Coverage Limit Amount To 1000",
            "context": "CarCoverage",
            "targetPath": "limitAmount",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "limitAmount",
                "contextName": "CarCoverage",
                "contextId": "-1",
                "newValue": 1000
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "Policy:-1:totalLimit": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Policy",
        "fieldName": "totalLimit",
        "fieldPath": "transactionDetails.totalLimit"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "Default Total Limit To Car Coverage Limit Amount",
            "context": "Policy",
            "targetPath": "totalLimit",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "transactionDetails.totalLimit",
                "contextName": "Policy",
                "contextId": "-1",
                "newValue": 1000
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    }
  }
]


kraken.engine.sanity.check.EngineSanityRuleOrderTest.shouldValidateEvaluateRulesInOrder=[
  {
    "BillingAddress:-1:countryCd": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "BillingAddress",
        "fieldName": "countryCd",
        "fieldPath": "countryCd"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "RULEORDER-BILLINGADDRESS",
            "context": "AddressInfo",
            "targetPath": "countryCd",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "countryCd",
                "contextName": "BillingAddress",
                "contextId": "-1",
                "newValue": "Address"
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "COLLCoverage:-1:code": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "COLLCoverage",
        "fieldName": "code",
        "fieldPath": "code"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "RULEORDER-COLL",
            "context": "COLLCoverage",
            "targetPath": "code",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "code",
                "contextName": "COLLCoverage",
                "contextId": "-1",
                "newValue": "PartyPartyAddress"
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "CreditCardInfo:-1:cardType": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "CreditCardInfo",
        "fieldName": "cardType",
        "fieldPath": "cardType"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "RULEORDER-CREDITCARD",
            "context": "CreditCardInfo",
            "targetPath": "cardType",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "cardType",
                "contextName": "CreditCardInfo",
                "contextId": "-1",
                "newValue": "PartyAddress"
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "Insured:-1:name": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Insured",
        "fieldName": "name",
        "fieldPath": "name"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "RULEORDER-INSURED",
            "context": "Insured",
            "targetPath": "name",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "name",
                "contextName": "Insured",
                "contextId": "-1",
                "newValue": "Party"
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "Party:-1:relationToPrimaryInsured": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Party",
        "fieldName": "relationToPrimaryInsured",
        "fieldPath": "relationToPrimaryInsured"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "RULEORDER-PARTY",
            "context": "Party",
            "targetPath": "relationToPrimaryInsured",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "relationToPrimaryInsured",
                "contextName": "Party",
                "contextId": "-1",
                "newValue": "Party"
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "Policy:-1:policyNumber": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Policy",
        "fieldName": "policyNumber",
        "fieldPath": "policyNumber"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "RULEORDER-POLICY",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "policyNumber",
                "contextName": "Policy",
                "contextId": "-1",
                "newValue": "PartyPartyAddress"
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "Vehicle:-1:model": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Vehicle",
        "fieldName": "model",
        "fieldPath": "model"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "RULEORDER-VEHICLE",
            "context": "Vehicle",
            "targetPath": "model",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "model",
                "contextName": "Vehicle",
                "contextId": "-1",
                "newValue": "PartyPartyAddress"
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    }
  }
]


kraken.engine.sanity.check.EngineSanityRuleOrderTest.shouldValidateRulesWithComplexFieldExpressionsInOrder=[
  {
    "BillingAddress:-1:city": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "BillingAddress",
        "fieldName": "city",
        "fieldPath": "city"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "RuleOrderWithComplexField_AddressInfo_city",
            "context": "AddressInfo",
            "targetPath": "city",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "city",
                "contextName": "BillingAddress",
                "contextId": "-1",
                "newValue": "San Diego"
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "BillingAddress:-1:doNotSolicit": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "BillingAddress",
        "fieldName": "doNotSolicit",
        "fieldPath": "doNotSolicit"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "RuleOrderWithComplexField_AddressInfo_doNotSolicit",
            "context": "AddressInfo",
            "targetPath": "doNotSolicit",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "doNotSolicit",
                "contextName": "BillingAddress",
                "contextId": "-1",
                "newValue": true
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "PartyRole:-1:limit": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "PartyRole",
        "fieldName": "limit",
        "fieldPath": "limit"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "RuleOrderWithComplexField_PartyRole_limit",
            "context": "PartyRole",
            "targetPath": "limit",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "limit",
                "contextName": "PartyRole",
                "contextId": "-1",
                "newValue": 100
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "Policy:-1:policyNumber": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Policy",
        "fieldName": "policyNumber",
        "fieldPath": "policyNumber"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "RuleOrderWithComplexField_Policy_policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "policyNumber",
                "contextName": "Policy",
                "contextId": "-1",
                "newValue": "doNotSolicit is true"
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "RRCoverage:-1:deductibleAmount": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "RRCoverage",
        "fieldName": "deductibleAmount",
        "fieldPath": "deductibleAmount"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "RuleOrderWithComplexField_AddressInfo_street",
            "context": "RRCoverage",
            "targetPath": "deductibleAmount",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "deductibleAmount",
                "contextName": "RRCoverage",
                "contextId": "-1",
                "newValue": 10
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "RRCoverage:-1:limitAmount": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "RRCoverage",
        "fieldName": "limitAmount",
        "fieldPath": "limitAmount"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "RuleOrderWithComplexField_RRCoverage_limitAmount",
            "context": "RRCoverage",
            "targetPath": "limitAmount",
            "payloadType": "DEFAULT"
          },
          "payloadResult": {
            "events": [
              {
                "attributeTarget": "limitAmount",
                "contextName": "RRCoverage",
                "contextId": "-1",
                "newValue": 10
              }
            ]
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    }
  }
]